edx-solutions/api-integration
# Mobile proficiency API with OAuth support

This implements API with OAauth support for student grades. It also adds user's course grade and cohort average grade to the API response.

**JIRA Ticket**: [OC-2886](https://tasks.opencraft.com/browse/OC-2886)

**Issues**: 
1. I am opening this PR although it is not fully tested, as I'm having some issues with my setup and there are parts of the code I am unsure about, so I need some feedback and help with testing.
2. I have added the code for retrieving the grades and calculating the average in the view, but there are probably more suitable places. I wasn't sure which is the preferred way, this, a mixin in users/views.py or persisted grades so I left the basic functionality to get some feedback from the reviewers and adjust the code accordingly.
3. I used a permission mixin initially, but the 403's and 404's are mixed in the same if statements
4. The api-integration tests take around 45 minutes to run (although I have a fast machine), 90% of the time being spent on setting up the databases, so it is quite time-consuming to test my changes and attempted bugfixes. I tried getting help to speed it up, but I didn't have any luck. Unfortunately I haven't been able to finish the unit tests for this reason, but I am committing them so someone can give me a hand (also I am going to be away until the afternoon, so it's good for someone to have a look before). Hopefully I'm the only one with this slow tests issue. 

**Testing instructions**:

1. Set up OAuth client in LMS admin if you don't already have one set up.
2. Obtain OAuth access token for your test user: curl -X POST -d "client_id={client_id}&client_secret={client_secret}&grant_type=password&username={username}&password={password}" https://lms.apros.local/oauth2/access_token/
3. Verify that you can access the api with your OAuth access token: curl 'http://lms.mcka.local/api/server/mobile/v1/users/courses/grades?username={username}&course_id={course_id} -H 'Authorization: Bearer {access_token}' -v
4. Make sure the user grades are returned and that their cohort average grade is calculated correctly.
5. Check that no inconsistencies can occur when staff users retrieve other user's grades

**Reviewers**:
